### PR TITLE
build(Dockerfiles): Refactor dockerfiles for backend and worker

### PR DESCRIPTION
**Pull request by stanleynguyen**
## Problem

[Ticket](https://www.notion.so/opengov/Optimizing-Dockerfiles-af57454eef2d4e90bc5e59233a32bfef?pvs=4)

## Solution

Switched to multi-step build and add cache to docker build on GHA:
- Reduced the image size by 20%
- Docker build time down from 2m47s (without cache) to 10s (with full cache) or more realistically 1m30s (without differing application code)

## Deployment Checklist

_Any tasks that need to be done on the environment before releasing of this PR (e.g. Setting environment variables, running migrations, etc). If there's no task to be done, put "N/A"_

- [ ] N/A

### CODE REVIEW

File: backend/Dockerfile [hunk 1]
```diff
@@ -1,4 +1,19 @@
...
+FROM public.ecr.aws/docker/library/node:18-alpine as builder
+WORKDIR /usr/home/postmangovsg
+
+COPY shared ../shared
+RUN cd ../shared && npm ci
+
+COPY ./package* ./
+RUN npm ci
+
+COPY tsconfig.* ./
```
Comment (tan-zx) [inline backend/Dockerfile:10]: trivial, but `tsconfig` should come before the others, since it's less likely to change?

File: worker/Dockerfile [hunk 2]
```diff
@@ -1,4 +1,20 @@
...
+FROM public.ecr.aws/docker/library/node:18-alpine as builder
+WORKDIR /usr/home/postmangovsg
+
+COPY shared ../shared
+RUN cd ../shared && npm ci
+
+COPY ./package* ./
+RUN npm ci
+
+COPY tsconfig.json ./
```
Comment (tan-zx) [inline worker/Dockerfile:10]: > trivial, but `tsconfig` should come before the others, since it's less likely to change?

same as prev

**Review by tan-zx — COMMENTED** [review]
honestly, I don't really understand Docker, so can't really review this PR — have left some best effort comments below

look forward to discussing irl

**Review by richard-yak — COMMENTED** [review]
same here, im not too familiar with docker in general. Perhaps @hanu-ogp can consider weighing in?

File: backend/Dockerfile [hunk 3]
```diff
@@ -13,24 +28,25 @@ RUN aws configure set default.region ap-southeast-1
...
+EXPOSE 4000
+COPY ./docker-entrypoint.sh ./
+RUN ["chmod", "+x", "docker-entrypoint.sh"]
 
-COPY tsconfig.json ./
-COPY tsconfig.build.json ./
-COPY openapi.yaml ./
-COPY src ./src
+COPY --from=builder /usr/home/shared/package* ../shared/
+RUN cd ../shared && npm install
```
Comment (hanu-ogp) [inline backend/Dockerfile:41]: Do we need npm install again if the folder is copied from the builder step ?

File: backend/Dockerfile [hunk 4]
```diff
@@ -13,24 +28,25 @@ RUN aws configure set default.region ap-southeast-1
 
 WORKDIR /usr/home/postmangovsg
 
-COPY shared ../shared
-RUN cd ../shared && npm ci
+# this means npm will only install production deps, and optimizations in the node
+# engine will take place as well
+ENV NODE_ENV=production
```
Comment (hanu-ogp) [inline backend/Dockerfile:33]: shouldnt this step be in the builder phase ?

File: worker/Dockerfile [hunk 5]
```diff
@@ -12,22 +28,24 @@ RUN aws configure set default.region ap-southeast-1
...
-RUN npm ci
+# expose port and copy these files first as they're unlikely to change
+EXPOSE 4000
+COPY ./docker-entrypoint.sh ./
+RUN ["chmod", "+x", "docker-entrypoint.sh"]
 
-COPY src ./src
-COPY tsconfig.json ./
+COPY --from=builder /usr/home/shared/package* ../shared/
+RUN cd ../shared && npm install
```
Comment (hanu-ogp) [inline worker/Dockerfile:41]: ditto as above

File: backend/Dockerfile (hunk 3 above)
Comment (stanleynguyen) [reply backend/Dockerfile:41]: Nope, we only copy package files

File: backend/Dockerfile (hunk 4 above)
Comment (stanleynguyen) [reply backend/Dockerfile:33]: Nope, we need devDependencies in the builder stage to build the app, plus we want this env var set in the final container as well

File: worker/Dockerfile (hunk 5 above)
Comment (stanleynguyen) [reply worker/Dockerfile:41]: Nope, we only copy package files

**Review by tan-zx — APPROVED** [review]
lgtm thanks for the changes 

hope to learn more about multi stage builds and layer caching when we meet irl

File: backend/Dockerfile [hunk 6]
```diff
@@ -1,4 +1,19 @@
...
+COPY tsconfig.* ./
+COPY ./package* ./
+RUN npm ci
+
+COPY src ./src
+
+RUN npm run build
+RUN cd ../shared && npm run postbuild
+
+FROM public.ecr.aws/docker/library/node:18-alpine
```
Comment (tan-zx) [inline backend/Dockerfile:16]: swap order with comment

File: backend/Dockerfile (hunk 4 above)
Comment (tan-zx) [inline backend/Dockerfile:33]: two purposes for setting this
1. you will only download dependencies and not devDependencies
2. you will run in optimised mode

File: worker/Dockerfile [hunk 7]
```diff
@@ -1,4 +1,20 @@
...
+COPY ./package* ./
+RUN npm ci
+
+COPY src ./src
+
+RUN npm run build
+RUN cd ../shared && npm run postbuild
+
+FROM public.ecr.aws/docker/library/node:18-alpine
+# postman public ecr
```
Comment (tan-zx) [inline worker/Dockerfile:17]: swap comment order

**Pull request merged**